util/rate_limiter.cc

```diff
@@ -139,148 +139,105 @@ void GenericRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
+  // granting will typically be done by a request at the front of its queue.
+  // This is important because one wakeup is saved when a granted request
+  // happens to be in an already running thread.
+  //
+  // Note this nice property is not guaranteed in a few cases, however.
+  //
+  // - No request may be granted.
+  // - Requests from a different queue may be granted.
+  // - (2) may be run by a non-front request thread causing it to not be granted
+  //   even if some requests in that same queue are granted. It can happen for a
+  //   couple (unlikely) reasons.
+  //    - A new request may sneak in and grab the lock at the refill time,
+  //      before the thread finishing (1) can wake up and grab it.
+  //    - A new request may sneak in and grab the lock and execute (1) before
+  //      (2)'s chosen candidate can wake up and grab the lock. Then that non-
+  //      front request thread performing (1) can carry over to perform (2).
   do {
-    bool timedout = false;
-
-    // Leader election:
-    //  Leader request's duty:
-    //  (1) Waiting for the next refill time;
-    //  (2) Refilling the bytes and granting requests.
-    //
-    //  If the following three conditions are all true for a request,
-    //  then the request is selected as a leader:
-    //  (1) The request thread acquired the request_mutex_ and is running;
-    //  (2) There is currently no leader;
-    //  (3) The request sits at the front of a queue.
-    //
-    //  If not selected as a leader, the request thread will wait
-    //  for one of the following signals to wake up and
-    //  compete for the request_mutex_:
-    //  (1) Signal from the previous leader to exit since its requested bytes
-    //      are fully granted;
-    //  (2) Signal from the previous leader to particpate in next-round
-    //      leader election;
-    //  (3) Signal from rate limiter's destructor as part of the clean-up.
-    //
-    //  Therefore, a leader request can only be one of the following types:
-    //  (1) a new incoming request placed at the front of a queue;
-    //  (2) a previous leader request whose quota has not been not fully
-    //      granted yet due to its lower priority, hence still at
-    //      the front of a queue;
-    //  (3) a waiting request at the front of a queue, which got
-    //      signaled by the previous leader to participate in leader election.
-    if (leader_ == nullptr &&
-        ((!queue_[Env::IO_HIGH].empty() &&
-            &r == queue_[Env::IO_HIGH].front()) ||
-         (!queue_[Env::IO_LOW].empty() &&
-            &r == queue_[Env::IO_LOW].front()))) {
-      leader_ = &r;
-
-      int64_t delta = next_refill_us_ - NowMicrosMonotonic();
-      delta = delta > 0 ? delta : 0;
-      if (delta == 0) {
-        timedout = true;
+    int64_t time_until_refill_us = next_refill_us_ - NowMicrosMonotonic();
+    if (time_until_refill_us > 0) {
+      if (waiting_) {
```

Comment:
Suggestion (with bias): 
I have a bias toward verbose variable naming and I myself might name "waiting_" something like "exists_waiting_thread_for_refill_time_"

Reply:
Hm let me try something like  `wait_for_refill_pending_`

Reply:
Good with me too!